test/performance/odp_timer_stress.c

```diff
@@ -0,0 +1,1119 @@
+	thr_common.instance = config->odp_instance;
+	thr_common.cpumask = &config->worker_mask;
+
+	for (uint32_t i = 0; i < config->opts.num_workers; ++i) {
+		thr_param = &thr_params[i];
+		worker = &config->worker_config[i];
+		odph_thread_param_init(thr_param);
+		thr_param->start = start_fn;
+		thr_param->thr_type = ODP_THREAD_WORKER;
+		thr_param->arg = worker;
+	}
+
+	if ((uint32_t)odph_thread_create(config->thread_tbl, &thr_common, thr_params,
+					 config->opts.num_workers) != config->opts.num_workers) {
+		ODPH_ERR("Error configuring worker threads\n");
+		return false;
+	}
+
+	return true;
+}
+
+static odp_bool_t setup_test(prog_config_t *config)
+{
+	return setup_config(config) && setup_workers(config);
+}
+
+static void run_control(prog_config_t *config)
+{
+	const uint32_t time_sec = config->opts.time_sec;
+	odp_atomic_u32_t *is_running = &config->is_running;
+
+	odp_barrier_wait(&config->init_barrier);
+
+	if (time_sec > 0U) {
+		sleep(time_sec);
+		odp_atomic_store_u32(is_running, 0U);
+	} else {
+		while (odp_atomic_load_u32(is_running))
+			sleep(1U);
+	}
+
+	odp_barrier_wait(&config->term_barrier);
+	(void)odph_thread_join(config->thread_tbl, config->opts.num_workers);
+}
+
+static void print_humanised(uint64_t value)
+{
+	if (value > GIGAS)
+		printf("%.2f GOPS\n", (double)value / GIGAS);
+	else if (value > MEGAS)
+		printf("%.2f MOPS\n", (double)value / MEGAS);
+	else if (value > KILOS)
+		printf("%.2f kOPS\n", (double)value / KILOS);
+	else
+		printf("%" PRIu64 " OPS\n", value);
+}
+
+static void print_stats(const prog_config_t *config)
+{
+	const stats_t *stats;
+	const opts_t *opts = &config->opts;
+
+	printf("" PROG_NAME " done:\n"
+	       "=====================\n\n"
+	       "  mode:             %s\n"
+	       "  clock source:     %d\n"
+	       "  resolution:       %" PRIu64 " ns\n"
+	       "  number of timers: %u (%s)\n\n", opts->mode == SINGLE_SHOT ? "single shot" :
+							opts->mode == PERIODIC ?
+							    "periodic" : "single shot with cancel",
+	       opts->clk_src, opts->res_ns, opts->num_tmr,
+	       opts->policy == SHARED_TMR ? "shared" : "private");
+
+	for (uint32_t i = 0U; i < config->opts.num_workers; ++i) {
+		stats = &config->worker_config[i].stats;
+
+		printf("  worker %u:\n"
+		       "    %s%" PRIu64 "\n", i,
+		       opts->mode == SINGLE_SHOT || opts->mode == PERIODIC ?
+			"number of timeouts: " : "number of cancels:  ",
+		       stats->num_tmo);
+
+		if (opts->mode == SINGLE_SHOT || opts->mode == CANCEL) {
+			if (opts->mode == CANCEL)
+				printf("    number of misses:   %" PRIu64 "\n", stats->num_miss);
+
+			printf("    number of retries:  %" PRIu64 "\n"
+			       "    max start mul:      %" PRIu64 "\n", stats->num_retry,
+			       stats->max_mul);
+		}
+
+		printf("    runtime:            %" PRIu64 " ns\n"
+		       "    rate:               ", stats->tot_tm);
+		print_humanised(stats->num_tmo / ((double)stats->tot_tm / ODP_TIME_SEC_IN_NS));
+		printf("\n");
+	}
+
```

Comment:
It could be useful to print also combined total stats for all workers (e.g. total rate).

Reply:
"total" section in V5.